### 1. Decouverte du dataset

- colonne : year, month, intent, police (impliqué ou non), sex, age, race, hispanic, place, eduction)

#### Instructions
- Lire le dataset "guns.csv" dans une liste data
- Afficher les 5 premières valeurs
- Extraire l'entête et l'assigner à la variable headers
- Supprimer l'en-tête de data
- Afficher l'en-tête
- Afficher de nouveau les 5 premières valeurs sans l'en-tête

In [3]:
import csv
f = open("guns.csv", 'r',encoding="utf-8")
data = list(csv.reader(f))

# Afficher les 5 premières valeurs
print(data[0:5])

# Extraire l'en-tête
headers = data[0] # ou headers = data[:1]

# Supprimer l'en-tête
data = data[1:]

# Afficher l'en-tête
print(headers)

# Afficher les 5 premières valeurs
print(data[0:5])

[['', 'year', 'month', 'intent', 'police', 'sex', 'age', 'race', 'hispanic', 'place', 'education'], ['1', '2012', '01', 'Suicide', '0', 'M', '34', 'Asian/Pacific Islander', '100', 'Home', '4'], ['2', '2012', '01', 'Suicide', '0', 'F', '21', 'White', '100', 'Street', '3'], ['3', '2012', '01', 'Suicide', '0', 'M', '60', 'White', '100', 'Other specified', '4'], ['4', '2012', '02', 'Suicide', '0', 'M', '64', 'White', '100', 'Home', '4']]
['', 'year', 'month', 'intent', 'police', 'sex', 'age', 'race', 'hispanic', 'place', 'education']
[['1', '2012', '01', 'Suicide', '0', 'M', '34', 'Asian/Pacific Islander', '100', 'Home', '4'], ['2', '2012', '01', 'Suicide', '0', 'F', '21', 'White', '100', 'Street', '3'], ['3', '2012', '01', 'Suicide', '0', 'M', '60', 'White', '100', 'Other specified', '4'], ['4', '2012', '02', 'Suicide', '0', 'M', '64', 'White', '100', 'Home', '4'], ['5', '2012', '02', 'Suicide', '0', 'M', '31', 'White', '100', 'Other specified', '2']]


### 2. Compter le nombre de décès par armes à feu aux USA par année
#### Instructions
- Utiliser une comprehension de listes pour extraire la donnée de la colonne year et assigner le résultat à la variable years
- Créer un dictionnaire vide qu'on nommera year_counts
- Compter le nombre de décès pour chaque année en parcourant la liste years
- Afficher year_counts

In [11]:
years = [y[1] for y in data]
year_counts = {}
for year in years:
    if year in year_counts:
        year_counts[year] += 1
    else:
        year_counts[year] = 1

print(year_counts)

{'2012': 33563, '2013': 33636, '2014': 33599}


### 3. Explorer  les décès par armes à feu par mois et année
#### Instructions
- Utiliser une compréhension de liste pour créer un object datetime pour chaque élément de la liste data affin d'extraire toutes les dates. Assigner le résultat à la variable dates.
    - N'oublier pas de convertir les valeurs de l'année et du mois avec int()
    - Utiliser les arguments year, month et day = 1 pour notre fonction datatime()
- Afficher les 5 premiers éléments de dates
- Compter le nombre d'élément par dates. Assigner le résultat à date_counts
- Afficher le résultat

In [24]:
import datetime
dates = [datetime.datetime(year = int(row[1]), month = int(row[2]), day = 1) for row in data]

print(dates[0:5])
date_counts = {}
for date in dates:
    if date not in date_counts:
        date_counts[date] = 0
    date_counts[date] += 1

print("\n")
print(date_counts)

[datetime.datetime(2012, 1, 1, 0, 0), datetime.datetime(2012, 1, 1, 0, 0), datetime.datetime(2012, 1, 1, 0, 0), datetime.datetime(2012, 2, 1, 0, 0), datetime.datetime(2012, 2, 1, 0, 0)]


{datetime.datetime(2012, 1, 1, 0, 0): 2758, datetime.datetime(2012, 2, 1, 0, 0): 2357, datetime.datetime(2012, 3, 1, 0, 0): 2743, datetime.datetime(2012, 4, 1, 0, 0): 2795, datetime.datetime(2012, 5, 1, 0, 0): 2999, datetime.datetime(2012, 6, 1, 0, 0): 2826, datetime.datetime(2012, 7, 1, 0, 0): 3026, datetime.datetime(2012, 8, 1, 0, 0): 2954, datetime.datetime(2012, 9, 1, 0, 0): 2852, datetime.datetime(2012, 10, 1, 0, 0): 2733, datetime.datetime(2012, 11, 1, 0, 0): 2729, datetime.datetime(2012, 12, 1, 0, 0): 2791, datetime.datetime(2013, 1, 1, 0, 0): 2864, datetime.datetime(2013, 2, 1, 0, 0): 2375, datetime.datetime(2013, 3, 1, 0, 0): 2862, datetime.datetime(2013, 4, 1, 0, 0): 2798, datetime.datetime(2013, 5, 1, 0, 0): 2806, datetime.datetime(2013, 6, 1, 0, 0): 2920, datetime.datetime(2013, 7, 1, 0, 0

### 4. Explorer les décès par arme à feu par origine et sexe
#### Instructions
- Compter le nombre de décès en fonction du sexe, assigner à la variable sex_counts
- Compter le nombre de décès en fonction de l'origine, assigner le résultat à race_counts
- Afficher sex_counts et race_counts

#### Nombre de décès par sexe

In [25]:
gender = [row[5] for row in data]
print(gender[:5])

['M', 'F', 'M', 'M', 'M']


In [26]:
sex_counts = {}
for g in gender:
    if g not in sex_counts:
        sex_counts[g] = 0
    sex_counts[g] +=1
print(sex_counts)

{'M': 86349, 'F': 14449}


#### Nombre de décès par race

In [29]:
races = [row[7] for row in data]
print(races[:5])

['Asian/Pacific Islander', 'White', 'White', 'White', 'White']


In [33]:
race_counts = {}
for r in races:
    if r not in race_counts:
        race_counts[r] = 0
    race_counts[r] += 1
print(race_counts)

{'Asian/Pacific Islander': 1326, 'White': 66237, 'Native American/Native Alaskan': 917, 'Black': 23296, 'Hispanic': 9022}


On peut dire que le nombre de décès par arme à feu est disproportionné en fonction si on est un homme et une femme. Cela toucherait plus les Hommes

On peut aussi dire que le nombre de décès par arme à feux est disprortionnée en fonction de la race mais on ne peut pas conclure car on ne connait pas le pourcentage démographique en fonction des races.

On peut aussi voir que les mois de juin à aout y a une légère augmentation du nombre de décès, il y a une saisonnalité, il y a un pic en été et un creux en hiver.

### Dataset population

In [34]:
f = open("census.csv", "r")
census = list(csv.reader(f))

print(census)

[['Id', 'Year', 'Id', 'Sex', 'Id', 'Hispanic Origin', 'Id', 'Id2', 'Geography', 'Total', 'Race Alone - White', 'Race Alone - Hispanic', 'Race Alone - Black or African American', 'Race Alone - American Indian and Alaska Native', 'Race Alone - Asian', 'Race Alone - Native Hawaiian and Other Pacific Islander', 'Two or More Races'], ['cen42010', 'April 1, 2010 Census', 'totsex', 'Both Sexes', 'tothisp', 'Total', '0100000US', '', 'United States', '308745538', '197318956', '44618105', '40250635', '3739506', '15159516', '674625', '6984195']]


### 6. Ratio de décès par armes à feu en fonction de l'origine

In [35]:
race_counts

{'Asian/Pacific Islander': 1326,
 'White': 66237,
 'Native American/Native Alaskan': 917,
 'Black': 23296,
 'Hispanic': 9022}

In [37]:
rate = 66237 / 197318956 * 100000
rate

33.56849303419181

- Asian/Pacific Islander = Race Alone - Asian plus Race Alone + Native Hawaïen and Other Pacific Islander
- Black = Race Alone - Black or African American
- Hispanic = Race Alone - Hispanic
- Native American/Native Alaskan = Race Alone - American Indian and Alaska Native
- Withe = Race Alone - White

#### Instructions
- Créer manuellement un dictionnaire nommé mapping qui contient les 5 origine avec le nombre d'habitants correspondant au dataset census
    - Les clés du dictionnaire seront  les 5 ci-dessus
    - Pour les Asian/Pacific, il faudra ajouter Race Alone - Asian - Native Hawaïian and Other Pacific Islander
- Créer un dictionnaire vide ratio
- Parcourir chaque éléments de race_counts précédemment créé
    - Pour chaque élément diviser la valeur de race_counts par la valeur associée au dictionnaire mapping
    - Multiplier par 100000
    - Assigner le résultat à la clé correspondant du dictionnaire ratio
- Afficher le ratio

#### Exploration des ratios de décès en fonction de l'origine

In [47]:
mapping = {
    'Asian/Pacific Islander': int(census[1][14]) + int(census[1][15]),
    'White': int(census[1][10]),
    'Native American/Native Alaskan': int(census[1][13]),
    'Black': int(census[1][12]),
    'Hispanic': int(census[1][11])
}  
print(mapping)


{'Asian/Pacific Islander': 15834141, 'White': 197318956, 'Native American/Native Alaskan': 3739506, 'Black': 40250635, 'Hispanic': 44618105}


In [48]:
ratio = {}
for race in race_counts:
    rate = race_counts[race] / mapping[race] * 100000
    ratio[race] = rate

print(ratio)

{'Asian/Pacific Islander': 8.374309664161762, 'White': 33.56849303419181, 'Native American/Native Alaskan': 24.521955573811088, 'Black': 57.8773477735196, 'Hispanic': 20.220491210910907}


### 7. Filtrer par homicide
#### Instructions
- Extraire la colonne intent en utilisant la compréhension de listes et assigner le résultats à la variable intents
- Extraire la colonne race en utilisant aussi une compréhension de listes et assigner le résultats à la variable races
- Créer un dictionnaire vide qu'on nommera homicide_counts
- Utiliser la fonction enumerate() pour parcourir chaque élément de la liste races et compter le nombre d'homicide par origine
- Réaliser la même procédure que précédemment pour calculer le ratio, assigner le résultat au dictionnaire homicide_ratio
- Afficher le résultat et conclure

In [50]:
intents = [row[3] for row in data]
races = [row[7] for row in data]

In [55]:
homicide_counts = {}
for i, race in enumerate(races):
    if race not in homicide_counts:
        homicide_counts[race] = 0
    if intents[i] == "Homicide":
        homicide_counts[race] +=1
    
print(homicide_counts)

{'Asian/Pacific Islander': 559, 'White': 9147, 'Native American/Native Alaskan': 326, 'Black': 19510, 'Hispanic': 5634}


In [56]:
homicide_ratio = {}
for race in homicide_counts:
    rate = homicide_counts[race] / mapping[race] * 100000
    homicide_ratio[race] = rate

print(homicide_ratio)

{'Asian/Pacific Islander': 3.530346230970155, 'White': 4.6356417981453335, 'Native American/Native Alaskan': 8.717729026240365, 'Black': 48.471284987180944, 'Hispanic': 12.627161104219914}


### Conclusion
on peut voir avec le ratio que les homicides affectent disproportionnellement plus les personnes de race Black  (48.47) et à moindre mesure les hispanics

### Nouvelles Pistes
- lien entre les mois et le ratio homicide
- lien entre le ratio homicide et sexe
- lien entre le ratio et les autres causes de décès (ex : suicide)
- ratio des décés par location et niveau d'éducation